db/cron.c

```diff
@@ -0,0 +1,614 @@
+        pthread_cond_timedwait(&sched->cond, &sched->mtx, &now);
+    }
+}
+
+void cron_unlock(cron_sched_t *sched)
+{
+    Pthread_mutex_unlock(&sched->mtx);
+}
+
+int cron_timedwait(cron_sched_t *sched, struct timespec *ts)
+{
+    return pthread_cond_timedwait(&sched->cond, &sched->mtx, ts);
+}
+
+const char *cron_type_to_name(enum cron_type type)
+{
+    switch (type) {
+    case CRON_TIMEPART:
+        return "WALLTIME";
+    case CRON_LOGICAL:
+        return "LOGICAL";
+    default:
+        return "UNKNOWN";
+    }
+}
+
+/**
+ * Return specific scheduler implementation
+ *
+ */
+sched_if_t *cron_impl(cron_sched_t *sched)
+{
+    return &sched->impl;
+}
+
+/***************** CRON SYSTABLE IMPLEMENTATION *******************/
+
+int cron_systable_schedulers_collect(void **data, int *nrecords)
+{
+    systable_cron_scheds_t *arr = NULL, *temparr;
+    cron_sched_t *sched;
+    int narr = 0;
+    int nsize = 0;
+    int rc;
+
+    Pthread_rwlock_rdlock(&crons.rwlock);
+    LISTC_FOR_EACH(&crons.scheds, sched, lnk)
+    {
+        if (narr >= nsize) {
+            nsize += 10;
+            temparr = realloc(arr, sizeof(systable_cron_scheds_t) * nsize);
+            if (!temparr) {
+                logmsg(LOGMSG_ERROR, "%s OOM %lu!\n", __func__,
+                       sizeof(systable_cron_scheds_t) * nsize);
+                cron_systable_schedulers_free(arr, narr);
+                narr = 0;
+                rc = -1;
+                goto done;
+            }
+            arr = temparr;
+        }
+        cron_lock(sched);
+        arr[narr].name = strdup(sched->impl.name);
+        arr[narr].type = strdup(cron_type_to_name(sched->impl.type));
+        arr[narr].running = sched->running;
+        arr[narr].nevents = sched->events.count;
+        arr[narr].description = sched->impl.describe
+                                    ? sched->impl.describe(&sched->impl)
+                                    : strdup("");
+        narr++;
+        cron_unlock(sched);
+    }
+done:
+    Pthread_rwlock_unlock(&crons.rwlock);
+    *data = arr;
+    *nrecords = narr;
+    return rc;
+}
+
+void cron_systable_schedulers_free(void *arr, int nrecords)
+{
+    systable_cron_scheds_t *parr = (systable_cron_scheds_t *)arr;
+    int i;
+
+    for (i = 0; i < nrecords; i++) {
+        if (parr[i].name)
+            free(parr[i].name);
+        if (parr[i].type)
+            free(parr[i].type);
+        if (parr[i].description)
+            free(parr[i].description);
+    }
+    free(arr);
+}
+
+int cron_systable_sched_events_collect(cron_sched_t *sched,
+                                       systable_cron_events_t **parr,
+                                       int *nrecords, int *pnsize)
+{
+    systable_cron_events_t *arr = *parr, *temparr;
+    int narr = *nrecords;
+    cron_event_t *event;
+    int nsize = *pnsize;
+    int rc = 0;
+    uuidstr_t us;
+
+    cron_lock(sched);
+    LISTC_FOR_EACH(&sched->events, event, lnk)
+    {
+        if (narr >= nsize) {
```

Comment:
Same as above, sched->events.count

Reply:
We don't block schedulers here while we collect events, so we don't know ahead of time how long the list it will be.